browser/src/control/Control.SidebarBase.ts

```diff
@@ -0,0 +1,161 @@
+		map: any,
+		options: SidebarOptions = {
+			animSpeed: 1000,
+		} /* Default speed: to be used on load */,
+		type: SidebarType,
+	) {
+		this.type = type;
+		this.options = options;
+		this.onAdd(map);
+	}
+
+	onAdd(map: ReturnType<typeof L.map>) {
+		this.map = map;
+
+		app.events.on('resize', this.onResize.bind(this));
+
+		this.builder = new L.control.jsDialogBuilder({
+			mobileWizard: this,
+			map: map,
+			cssClass: `jsdialog sidebar`, // use sidebar css for now, maybe have seperate css for navigator later
+		});
+		this.container = L.DomUtil.create(
+			'div',
+			`${this.type}-container`,
+			$(`#${this.type}-panel`).get(0),
+		);
+
+		this.map.on('jsdialogupdate', this.onJSUpdate, this);
+		this.map.on('jsdialogaction', this.onJSAction, this);
+	}
+
+	onRemove() {
+		this.map.off('jsdialogupdate', this.onJSUpdate, this);
+		this.map.off('jsdialogaction', this.onJSAction, this);
+	}
+
+	isVisible(): boolean {
+		return $(`#${this.type}-dock-wrapper`).hasClass('visible');
+	}
+
+	closeSidebar() {
+		$(`#${this.type}-dock-wrapper`).removeClass('visible');
+		this.map._onResize();
+
+		if (!this.map.editorHasFocus()) {
+			this.map.fire('editorgotfocus');
+			this.map.focus();
+		}
+
+		this.map.uiManager.setDocTypePref(`Show${this.type}`, false);
+	}
+
+	onJSUpdate(e: FireEvent) {
+		var data = e.data;
+
+		if (data.jsontype !== this.type) return;
+
+		if (!this.container) return;
+
+		if (!this.builder) return;
+
+		// reduce unwanted warnings in console
+		if (data.control.id === 'addonimage') {
+			window.app.console.log('Ignored update for control: ' + data.control.id);
+			return;
+		}
+
+		this.builder.updateWidget(this.container, data.control);
+	}
+
+	onJSAction(e: FireEvent) {
+		var data = e.data;
+
+		if (data.jsontype !== this.type) return;
+
+		if (!this.builder) return;
+
+		if (!this.container) return;
+
+		var innerData = data.data;
+		if (!innerData) return;
+
+		var controlId = innerData.control_id;
+
+		// Panels share the same name for main containers, do not execute actions for them
+		// if panel has to be shown or hidden, full update will appear
+		if (
+			controlId === 'contents' ||
+			controlId === 'Panel' ||
+			controlId === 'titlebar' ||
+			controlId === 'addonimage'
+		) {
+			window.app.console.log(
+				'Ignored action: ' +
+					innerData.action_type +
+					' for control: ' +
+					controlId,
+			);
+			return;
+		}
+
+		this.builder.executeAction(this.container, innerData);
+	}
+	markNavigatorTreeView(data: WidgetJSON): boolean {
+		if (!data) return false;
+
+		if (data.type === 'treelistbox') {
+			(data as TreeWidgetJSON).draggable = false;
+			return true;
+		}
+
+		for (const i in data.children) {
+			if (this.markNavigatorTreeView(data.children[i])) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+	onResize() {
+		var wrapper = document.getElementById(`${this.type}-dock-wrapper`);
+		wrapper.style.maxHeight =
+			document.getElementById('document-container').getBoundingClientRect()
```

Comment:
document container is not changed in time, can we store reference in the member variable on initialization? to not query DOM all the time.